ats_sms_operator/sender.py

```
@@ -0,0 +1,131 @@
+
+    return ''.join(chain(
+        (header.format(username=config.ATS_USERNAME, password=config.ATS_PASSWORD),),
+        (request.serialize_ats() for request in ats_requests),
+        (footer,),
+    ))
+
+
+def send_ats_requests(*ats_requests):
+    """
+    Performs the actual POST request with the given elementary ATS requests.
+    """
+    requests_xml = serialize_ats_requests(*ats_requests)
+    return requests.post(config.ATS_URL, data=requests_xml, headers={'Content-Type': 'text/xml'},
+                         slug='ATS SMS', related_objects=ats_requests)
+
+
+def parse_response_codes(xml):
+    """
+    Finds all <code> tags in the given XML and returns a mapping "uniq" -> "response code" for all SMS.
+    In case of an error, the error is logged.
+    """
+    soup = BeautifulSoup(xml, 'html.parser')
+    code_tags = soup.find_all('code')
+
+    LOGGER.warning(', '.join(
+        [config.ATS_STATES.get_label(c) if c in config.ATS_STATES.all else 'ATS returned an unknown state {}.'.format(c)
+         for c in [int(error_code.string) for error_code in code_tags if not error_code.attrs.get('uniq')]],
+    ))
+
+    return {int(code.attrs['uniq']): int(code.string) for code in code_tags if code.attrs.get('uniq')}
+
+
+def send_and_parse_response(*ats_requests):
+    """
+    Glue function to perform sending ATS requests and parsing the ATS server response in one go.
+    """
+    response = send_ats_requests(*ats_requests)
+    return parse_response_codes(response.text)
+
+
+def send_and_update_sms_states(*ats_requests):
+    """
+    Higher-level function performing serialization of ATS requests, parsing ATS server response and updating
+    SMS messages state according the received response.
+    """
+    for uniq, state in send_and_parse_response(*ats_requests).items():
+        sms = get_object_or_none(config.get_output_sms_model(), pk=uniq)
```

maybe I will raise SMSValidationError, if service returns sms with unknown uniq, because then probably something wrong happened